### ISSUE TITLE
Allow statements to use full type descriptors

### ISSUE THREAD

**Issue by jclark** [ballerina-platform/nballerina#412]
Currently we parse type descriptors in two ways:

- full type descriptors are allowed in type definitions
- "inline" type descriptors are allowed in statements

We need to improve this so that statements can contain full type descriptors.  This requires fixing both parsing and type resolution.

Part of the necessary work for type resolution was done in #339.

**Comment by jclark** [comment]
The fix for #411 fixes type resolution. The remaining part is parsing, which requires unbounded lookahead, for example:

```
(((x))).clone();
(((x))) n;
```

or

```
x[10] = 17;
x[10] foo;
```

**Comment by jclark** [comment]
Unbounded lookahead will be easier with #408.

**Comment by jclark** [comment]
With #454, we can do unbounded lookahead:

- make a new Tokenizer
- save the state of the current tokenizer
- restore the state into the new Tokenizer
- use the new Tokenizer to lookahead as much as needed

**Comment by jclark** [comment]
So what we have to figure out is, given this ability, what is the best way to figure out whether we have a local variable declaration or something else.  This requires looking at what follows the type descriptor (or what would be the type descriptor if it's a local variable declaration).  In the local variable declaration case it will be something that can start a binding-pattern (e.g. an identifier); in other cases, it will be something more operator-like (e.g. `.`).  The first task is to analyze the grammar and make this precise.

After that, we can figure out the strategy to actually implement it.

**Comment by jclark** [comment]
Note that for subset 8 (in #471), we really ought to be able to handle

m:T v = E;

**Comment by manuranga** [comment]
Meeting notes:

1) We should try to make do with left factoring when it's reasonable, instead of save/restore.
2) Qualified identifier should be done first to facilitate #438

**Comment by jclark** [comment]
Also consider a peek method on Tokenizer that would turn a FragCode into a SingleCharDelim|MultiCharDelim|Keyword|VariableLengthToken

**Comment by manuranga** [comment]
I plan to add a `string? modPrefix` to `VarRefExpr` and `TypeDescRef`.

**Comment by jclark** [comment]
Just call it `prefix`.

**Comment by jclark** [comment]
I think we're starting to blur the lines between this and #471.

This issue is really about allowing arbitrary type descriptors in local variable declaration statements (by doing lookahead).  So let's make a separate issue (#473) just for parsing of qualified identifier in variable references and type references.

**Comment by manuranga** [comment]
I have to think a bit more before finalizing, but at high level I thought of two approaches. Let's take the case where we see `(` at the beginning of a statement.

#### Approach 1 - single pass

```
Call startPrimaryExpr
If it is an error
    Parse type-desc
    If it's also an error
        return error, expected type-desc or expression
    Else
        Finish var decl
Else look at the current token
   If it's `.`
       Finish method-call-stmt
   If it's `[`
        Convert current expression to a type desc
        Finish parsePostfixTypeDesc
        Finish local-var-decl
   If it's identifier
        Finish local-var-decl
```

#### Approach 2 - look ahead

```
Save state
Set paranLevel to 1
Keep advancing until paranLevel is 0, while incrementing it on `(` and decrementing on `)`
If the current token is `.`
    Restore state and finish method-call-stmt
If it's `[` or identifier
    Restore state and finish local-var-decl    
```

Hope to discuss this in today meeting.

**Comment by jclark** [comment]
One point to bear in mind is that the grammar for type descriptors is a lot simpler than the grammar for expressions.  It is a bit easier to analyze what is impossible in a type-descriptor than in an expression.

Some things to look for:
- a built-in type such as `int`, `string`, `boolean` can only occur in an expression  if it immediately precedes `:` or if there has previously been one of the following tokens: `<` , `is`, `let`, `from`, `function`.
- some tokens cannot occur in a type descriptor at all e.g. `.`, `check`, `let`, `from`, `is`
- `<` can only occur after specific tokens in a type-descriptor e.g. `map`, `future`, `typedesc`, `error`; otherwise `<` indicates an expression

**Comment by jclark** [comment]
One way to approach this would be to have a state machine, with a stack of open brackets (`([{`), which you feed with token types (i.e. FixedToken|VariableTokenCode), and after each token, it says one of:

- keep going - not sure yet
- definitely local-var-decl
- definitely method-call
- error - can't be either

We can start relatively simple and add complexity as time goes on.

**Comment by jclark** [comment]
Further thoughts:

- recursive descent will be easier than a state machine
- probably not worth doing left factoring: starting with a parenthesis is not so common, and saving/restoring tokenizer state is not so expensive

So you have something like

```
function preparseParenTypeDesc(Tokenizer tok) returns boolean?|err:Syntax;
```

where
- boolean means we have definitely decided it is/isn't a type desc
- nil means we need to keep parsing
- err:Syntax is an error

After recursively calling preparseParenTypeDesc, it immediately returns any non-nil value.

**Comment by jclark** [comment]
I would add the expected close bracket as a parameter to the preparse function. Then you can do most of the work with a single function.

**Comment by manuranga** [comment]
Yesterday, I wrote a small ballerina program to [generate first non-terminal for each rule](https://gist.github.com/manuranga/c5a8efe59f06c458412798f571d6245c). It would be useful if this is extended to give next terminal, give a rule and current terminals, then even draw a railroad diagram. But I paused this yak-shaving today to work on the main task. I have created above pr with expected grammar and some tests cases.

Planing to add more test cases tomorrow and outline the impl of preparseParenTypeDesc.

**Comment by jclark** [comment]
This `[int][X]` is a nasty case, because `[X]` could be a binding pattern or a fixed length array.

So:

```
const X = 2;
[int][X] v = [[0], [1]];
```

or

```
[int] v = [1];
[int] [X] = v;
```

**Comment by manuranga** [comment]
Update:
I am in the process of extending the supported statement grammar to have parentheses around just the type ref and `()`, (in addition to union types defs). I think is a good way to stage, since it will require figuring out key features, which are:
- Recursively handling parentheses
- Keeping the code similar for `[` and `(`
- Distinguishing between `x[CONST]` as a type and ref. (Note that when we support parenthesis around type ref, we should also support `[` around identifiers for the code to be symmetric)

Examples of things to be covered under this stage:
```
()|int i = ();
(()|int) i = ();
(()) n = ();
(()).m();
(MyInt) i = 4;
(myInt).m();
(myArr)[1].m();
(myArr)[1 + 1].m();
```

**Comment by jclark** [comment]
I don't think it's useful to extend the subset grammar to do this: the point of this issue is to simplify the subset grammar by getting rid of the difference between inline and non-inline type descriptors.  Let's just get this done.  What are the cases that you cannot yet distinguish?

**Comment by jclark** [comment]
We don't need `x[CONST]` as a type descriptor for quite some time.

**Comment by jclark** [comment]
The code needs some work.

**Comment by jclark** [comment]
@manuranga I did a bit of work on preparsing. I have restructured it so that it will accomodate future evolution better. It gives errors in more cases. I didn't really understand your testing code.  Please have a look and maybe add a few more test cases. I added an early return for nested "." just to show how to accommodate this. You might want to add a few cases to the `match` statement.

**Comment by jclark** [comment]
There is still more work to do here.  We should be able to get rid of all the code for inline type desc parsing from parseTypeDesc (first 85 lines or so).  But we can't. Problem is parseErrorStmt using finishInlineTypeDesc in order to handle this (which is syntactically but not semantically valid):

```
error("foo").whatever();
```

**Comment by jclark** [comment]
To get rid of inline type desc from the subset grammar we also need to allow statements starting with `record`.

**Comment by jclark** [comment]
In the full grammar, there are difficult cases with error; `error X` could start `error X("foo").whatever()` or `error X = foo()`. So in general we need do a token save/restore like with `(`.

I will create a separate issue for dealing with simpler cases more straightforwardly.

**Comment by jclark** [comment]
Maybe we should just bite the bullet and do #545.

**Comment by jclark** [comment]
I'll take a shot at #545.

**Comment by jclark** [comment]
We should also have a couple of tests in the test suite for cases with parentheses to make sure everything is working together particularly this sort of thing

```
(int|string)[] v = [16];
```

which we could not do before.
